app/controllers/AssessmentController.scala

```diff
@@ -40,71 +42,73 @@ class AssessmentController @Inject() (
   getData: DataRetrievalAction,
   requireData: DataRequiredAction,
   formProvider: AssessmentFormProvider,
+  categorisationService: CategorisationService,
   val controllerComponents: MessagesControllerComponents,
   view: AssessmentView
 )(implicit ec: ExecutionContext)
     extends FrontendBaseController
     with I18nSupport {
 
-  def onPageLoad(mode: Mode, assessmentId: String): Action[AnyContent] =
-    (identify andThen getData andThen requireData) { implicit request =>
-      {
-        for {
-          categorisationInfo <- request.userAnswers.get(CategorisationQuery)
-          assessment         <- categorisationInfo.categoryAssessments.find(_.id == assessmentId)
-          assessmentIndex     = categorisationInfo.categoryAssessments.indexOf(assessment)
-        } yield {
-
-          val form = formProvider(assessment.exemptions.map(_.id))
+  def onPageLoad(mode: Mode, recordId: String, index: Int): Action[AnyContent] =
+    (identify andThen getData andThen requireData).async { implicit request =>
+      val categorisationResult = for {
+        userAnswersWithCategorisations <- categorisationService.requireCategorisation(request, recordId)
+        recordQuery                    <- Future.successful(userAnswersWithCategorisations.get(RecordCategorisationsQuery))
```

Comment:
you can use = rather than <- if you don't need to flatmap and just want to set a value in a for comprehension

Reply:
recordQuery =  userAnswersWithCategorisations.get(RecordCategorisationsQuery)

Reply:
That is cool, didn't know that, done :)

Reply:
That'll be why I was struggling with types in the for comprehension